### DISCUSSION TITLE
How to create a custom filter based on v6

### DISCUSSION THREAD

**Discussion by ericCHxia** [fabricjs/fabric.js#8863] [Q&A] [answered]
My project is based on TS and ES6, so I think v6 fits my needs. However, I didn't found any reference based on v6 about how to creare a custom filter. And when I try to write a new class that inherits from BaseFilter, I found some functions have parameter types that are not exported.

**Comment by asturur** [comment]
hi @ericCHxia , we recently converted filters to TS.
Can you show me your class extending BaseFilter so we can work through it together?

thanks

**Comment by asturur** [comment]
https://github.com/fabricjs/fabric.js/pull/8913

**Comment by AlvesJorge** [comment, marked as answer]
Reviving this thread if I may!
I'm also trying to create a filter with a custom shader in Fabric 6 (Thanks for the great work so far!).
I took the Grayscale class definition from source and just changed the shader to give a red tint instead of grayscale to test if it works
```javascript
class RedFilter extends fabric.filters.BaseFilter {
    getCacheKey() {
        return "".concat(this.type, "_").concat("red");
    }

    getFragmentSource() {
        return `
            precision highp float;
            uniform sampler2D uTexture;
            uniform int uMode;
            varying vec2 vTexCoord;
            void main() {
                vec4 col = texture2D(uTexture, vTexCoord);
                gl_FragColor = vec4(1.0, 0.0, 0.0, col.a);
            }
            `
    }

    getUniformLocations(gl, program) {
        return {
            uMode: gl.getUniformLocation(program, 'uMode')
        };
    }

    sendUniformData(gl, uniformLocations) {
        const mode = 1;
        gl.uniform1i(uniformLocations.uMode, mode);
    }

    isNeutralState() {
        return false;
    }
}

const redDefaultValues = {
    mode: 'red',
    mainParameter: 'mode'
};

function defineProperty(obj, key, value) {
    Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
}


defineProperty(RedFilter, "type", 'Grayscale');
defineProperty(RedFilter, "defaults", redDefaultValues);
classRegistry.setClass(RedFilter);
```
And it did work! I did have to copy `defineProperty` over from the source code, and I'm not sure this is the intended way to define the "type" property.

**Reply by ShaMan123** [reply]
`defineProperty` is the transpiled version of a static method/field

https://github.com/fabricjs/fabric.js/blob/dc4c208829616431f763854a6bf3ffdf09fb252f/src/filters/Grayscale.ts#L24-L26
